### PR TITLE
Add argument postGenerationCommand to ExternalLHEProducer and provide mergeLHE.py script [new]

### PR DESCRIPTION
**Pull request by colizz**
#### PR description:

(following #33615)
- Add an argument `postGenerationCommand` to ExternalLHEProducer module that can run a self-specified command right after LHE event generation (i.e. launched by the `ScriptName`).
- Add a general LHE merging script `mergeLHE.py` .
- Revert #33628 (switch `externalLHEProducer.generateConcurrently` to True), and resolve the concurrent ExternalLHEProducer interfacing with H7 issue raised in #33544.
  - Now we provide `postGenerationCommand` to merge LHE events after concurrent LHE event generation for the later use of H7GeneratorFilter.

Note that the idea to add `postGenerationCommand` to the cfg file is not designed for permanent use. It is a hack at the time being that handles the special case of Herwig7GeneratorFilter which can only be feed with an entire piece of LHE file as input. Meanwhile, the new script `mergeLHE.py` may be used as a general tool and also on other occasions.

#### PR validation:

Pass workflow 535.0 (switch `externalLHEProducer.generateConcurrently` to True).

Pass the local validation on the `mergeLHE.py` script which are more specific to the generator group, see in the following comment.

### CODE REVIEW

**Comment by colizz** [comment]
### Physics validation in generator group

The `mergeLHE.py` script is designed to be "a universal script that merges multiple LHE files for all possible conditions and in the most natural way".
A test routine for this mergeLHE.py is set up [here](https://github.com/colizz/mergelhe_validate), as mentioned in the script, to check if it has expected performance on different LHE merging conditions.

More specifically we examine the physics result of the following case for Powheg+H7 (matrix 535.0) and MG LO+H7 ([`JME-RunIISummer20UL16wmLHEGEN-00003`](https://cms-pdmv.cern.ch/mcm/requests?page=0&prepid=JME-RunIISummer20UL16wmLHEGEN-00003)) case, in three schemes respectively:

 - using original mode, i.e. set `externalLHEProducer.generateConcurrently` to False;
 - setting `generateConcurrently` to True and merge LHEs via command `postGenerationCommand = cms.untracked.vstring('mergeLHE.py', '-i', 'thread*/cmsgrid_final.lhe', '-o', 'cmsgrid_final.lhe'),` (i.e., default merging)
 - setting `generateConcurrently` to True and merge LHEs via command `postGenerationCommand = cms.untracked.vstring('mergeLHE.py', '-i', 'thread*/cmsgrid_final.lhe', '-o', 'cmsgrid_final.lhe', '--bypass-check'),` (i.e., trivial merging, which bypasses the consistent check and thus does not modify the header/init/event block for merged LHEs).

See physics results for [Powheg+H7](https://coli.web.cern.ch/coli/validation/mergeLHEforH7/pow) and [MG LO+H7](https://coli.web.cern.ch/coli/validation/mergeLHEforH7/mgh7).

Results from the three schemes match each other, except that in the MG LO+H7 case we see discrepancies in `PSWeight[1]` to `PSWeight[4]` for the "default merging" and "trivial merging" scheme.

Explanation: these `PSWeight`s come from `optionalWeights()` in H7 and appear only in MG LO+H7 routine due to the use of `FxFxHandler` in H7. The corresponding 1344 entries from `optionalWeights()` are summarised [here](https://coli.web.cern.ch/coli/validation/mergeLHEforH7/mgh7/optionalWeights.txt), and the first few entries are corresponding to the LHE clustering info. Note that the weights from an MG LO LHE (both the per-event weight XWGTUP and all <wgt> values) may vary according to the number of events produced, thus by merging them one should consider modifying the weights. The weights are correctly handled in our "default merging" case but not in "trivial merging", so one sees the discrepancies in `PSWeight`. **Overall, we can trust the results given by "default merging".**

However, please note these `PSWeight` values are raw values returned by H7 `FxFxHandler` and it is yet to be known whether we will use them in the analysis (should consult the H7 experts).

File: Configuration/Generator/python/TTbar_Pow_LHE_13TeV_cff.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
     outputFile = cms.string('cmsgrid_final.lhe'),
     scriptName = cms.FileInPath('GeneratorInterface/LHEInterface/data/run_generic_tarball_cvmfs.sh'),
     generateConcurrently = cms.untracked.bool(True),
+    postGenerationCommand = cms.untracked.vstring('mergeLHE.py', '-i', 'thread*/cmsgrid_final.lhe', '-o', 'cmsgrid_final.lhe'),
```
Comment (qliphy) [inline Configuration/Generator/python/TTbar_Pow_LHE_13TeV_cff.py:10]: What will happen if we run crab jobs with generateConcurrently set to True?

**Comment by qliphy** [comment]
please test workflow 535.0

File: Configuration/Generator/python/TTbar_Pow_LHE_13TeV_cff.py (hunk 1 above)
Comment (colizz) [reply Configuration/Generator/python/TTbar_Pow_LHE_13TeV_cff.py:10]: Thanks @qliphy . Indeed before the "mergeLHE.py" script is merged into cmssw, the crab job may fail - it is pretty weird to me but the fact is that crab cannot find the path of "mergeLHE.py" if it is in the local cmssw `src/`. But I checked that after the script is merged, the issue will be resolved.

**Comment by agrohsje** [comment]
Hi @colizz. Thanks for the update. I checked again sum/average different number of events. All worked well and is consistent with direct production. I just noticed a small point. The option is called : --input-paths but in fact it expects input files. Do you think we should adjust? In any case I am happy to sign. Let me know.

**Comment by colizz** [comment]
> Hi @colizz. Thanks for the update. I checked again sum/average different number of events. All worked well and is consistent with direct production. I just noticed a small point. The option is called : --input-paths but in fact it expects input files. Do you think we should adjust? In any case I am happy to sign. Let me know.

Thanks for the test! I agree with this - paths may imply also directories which is not for our case. 
I just rename them to --input-files & --output-file (the latter can also be a file path in this script version).

**Comment by agrohsje** [comment]
+generators

**Comment by qliphy** [comment]
@colizz   There is an CMSSW Code Rules error log:   https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ddf8b1/15754/codeRules/cmsCodeRule3.txt
Please check the following to avoid usage of catch ( ... )
cms-sw.github.io/cms_coding_rules.html
and https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEdmExceptionUse#cms_Exception

**Comment by colizz** [comment]
Thanks @qliphy , this seems to come from an earlier implementation
https://github.com/cms-sw/cmssw/blob/60697c8665538c3f88fbfd03f1502c3468266a52/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc#L274
that catches all exception from the external LHE launching script. I think this syntax can be accepted here, since the purpose is to rethrow the exception coming from any thread
https://github.com/cms-sw/cmssw/blob/60697c8665538c3f88fbfd03f1502c3468266a52/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc#L286
What do you think? Or maybe we can add an `exceptPaths` to https://github.com/cms-sw/cmssw/blob/master/Utilities/ReleaseScripts/python/cmsCodeRules/config.py to skip the check.

**Comment by qliphy** [comment]
+1

**Pull request merged**